File: Data Structures & Algorithms/maximum-subarray/submission-0.java
class Solution {
    public int maxSubArray(int[] nums) {
        if (nums.length == 1) return nums[0];

        int [] sums = new int[nums.length];
        sums[0] = nums[0];
        for (int i = 1; i < nums.length; i++) {
            sums[i] = Math.max(sums[i - 1] + nums[i], nums[i]);
        }
        int max = sums[0];
        for (int i = 1; i < sums.length; i++) {
            max = Math.max(max, sums[i]);
        }


        return max;
    }
}

File: Data Structures & Algorithms/house-robber/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    rob(nums) {
            const cache = new Int32Array(nums.length).fill(-1);
        cache[0] = nums[0];
        cache[1] = nums[1];
        for (let i = 2; i < nums.length; i++) {

            let index = i - 2;
            let max = index;
            while (index >= 0) {
                index--;
                if (cache[index] > cache[max]) max = index;
            }

            cache[i] = nums[i] + cache[max];
        }

        let max = 0;

        for (const n of cache) {
            if (n > max) max = n;
        }

        return max;
        
    }
}
